dist/gcc-compatible/Hacl_Streaming_HMAC.h

```diff
@@ -0,0 +1,119 @@
+/* MIT License
+ *
+ * Copyright (c) 2016-2022 INRIA, CMU and Microsoft Corporation
+ * Copyright (c) 2022-2023 HACL* Contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+
+#ifndef __Hacl_Streaming_HMAC_H
+#define __Hacl_Streaming_HMAC_H
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+#include <string.h>
+#include "krml/internal/types.h"
+#include "krml/lowstar_endianness.h"
+#include "krml/internal/target.h"
+
+#include "Hacl_Streaming_Types.h"
+#include "Hacl_Krmllib.h"
+#include "Hacl_Hash_SHA3.h"
+#include "Hacl_Hash_SHA2.h"
+#include "Hacl_Hash_Blake2s_Simd128.h"
+#include "Hacl_Hash_Blake2s.h"
+#include "Hacl_Hash_Blake2b_Simd256.h"
+#include "Hacl_Hash_Blake2b.h"
+
+#define Hacl_Agile_Hash_MD5 0
+#define Hacl_Agile_Hash_SHA1 1
+#define Hacl_Agile_Hash_SHA2_224 2
+#define Hacl_Agile_Hash_SHA2_256 3
+#define Hacl_Agile_Hash_SHA2_384 4
+#define Hacl_Agile_Hash_SHA2_512 5
+#define Hacl_Agile_Hash_SHA3_224 6
+#define Hacl_Agile_Hash_SHA3_256 7
+#define Hacl_Agile_Hash_SHA3_384 8
+#define Hacl_Agile_Hash_SHA3_512 9
+#define Hacl_Agile_Hash_Blake2S_32 10
+#define Hacl_Agile_Hash_Blake2S_128 11
+#define Hacl_Agile_Hash_Blake2B_32 12
+#define Hacl_Agile_Hash_Blake2B_256 13
+
+typedef uint8_t Hacl_Agile_Hash_impl;
+
+typedef struct Hacl_Agile_Hash_state_s_s Hacl_Agile_Hash_state_s;
+
+typedef struct Hacl_Streaming_HMAC_Definitions_index_s
+{
+  Hacl_Agile_Hash_impl fst;
+  uint32_t snd;
+}
+Hacl_Streaming_HMAC_Definitions_index;
+
+typedef struct Hacl_Streaming_HMAC_Definitions_key_and_len_s
+{
+  uint8_t *fst;
```

Comment:
Quick implementation: do you keep the key in clear here? Since HMAC does

```
H(K' ^ opad || H(K' ^ ipad || text))
```

and the underlying hash function admits an update routine, can't you just store the state of the inner and outer layers instead?

Reply:
yes, I could, it would just require a little bit more work (and I'm off next week) -- do you think this is a blocker?

Reply:
And take your time. We have until May for it to be included in 3.14 and if it does not land in 3.14, it'll land in 3.15.

Reply:
Let's ask someone from CPython as well: @gpshead In the Python implementation of HMAC, we drop the key and only keep inner and outer states (well, we just keep the hash objects directly). Do you think this should be the same here ?

Reply:
I didn't realize I could perform that optimization, but I agree it's a nice one to have. The optimization would be a little bit more of verification work, I think, but not dramatically so, since most of the groundwork is already laid out. You would just have to incorporate into the invariant that the second hash state, upon initialization, has been fed only the key, and remains unchanged into the call to `digest`, upon which is receives the first hash.